In [23]:
import tensorflow as tf

In [24]:
import numpy as np

In [4]:
print(tf.__version__)

2.10.1


In [4]:
random_1 = tf.random.Generator.from_seed(42)
random_1 = random_1.normal(shape=(3,2))

In [5]:
random_1

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[-0.7565803 , -0.06854702],
       [ 0.07595026, -1.2573844 ],
       [-0.23193765, -1.8107855 ]], dtype=float32)>

In [19]:
random_2 = tf.random.Generator.from_seed(42)
random_2 = random_2.normal(shape =(3,2))
random_2 = tf.random.shuffle(random_2, seed=20)

In [20]:
random_2

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[-0.7565803 , -0.06854702],
       [ 0.07595026, -1.2573844 ],
       [-0.23193765, -1.8107855 ]], dtype=float32)>

In [11]:
random_1 == random_2

<tf.Tensor: shape=(3, 2), dtype=bool, numpy=
array([[ True,  True],
       [ True,  True],
       [ True,  True]])>

Operations that rely on a random seed actually derive it from two seeds: the global and operation-level seeds. This sets the global seed.
https://www.tensorflow.org/api_docs/python/tf/random/set_seed

In [24]:
not_shuffled = tf.constant([[10,7],
                            [3,4],
                            [2,5]
                           ])

In [25]:
not_shuffled

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]])>

In [60]:
tf.random.shuffle(not_shuffled, seed=42)
# if use only this,though seed has been set, but result still random

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 2,  5],
       [ 3,  4],
       [10,  7]])>

In [70]:
tf.random.set_seed(42)    # global level random seed
tf.random.shuffle(not_shuffled, seed=42)    # operation level random seed

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]])>

In [13]:
n = np.ones((2,2))

In [14]:
type(n)

numpy.ndarray

In [16]:
a = tf.ones([3,2,3,4])

In [17]:
type(a)

tensorflow.python.framework.ops.EagerTensor

In [87]:
a

<tf.Tensor: shape=(3, 2, 3, 4), dtype=float32, numpy=
array([[[[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]],

        [[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]]],


       [[[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]],

        [[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]]],


       [[[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]],

        [[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]]]], dtype=float32)>

In [88]:
a.ndim

4

In [19]:
tf.ones((2,2)) # bracket

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[1., 1.],
       [1., 1.]], dtype=float32)>

In [90]:
tf.ones([2,2])   # square bracket

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[1., 1.],
       [1., 1.]], dtype=float32)>

In [91]:
tf.zeros(shape=(3,4))   #bracket

<tf.Tensor: shape=(3, 4), dtype=float32, numpy=
array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]], dtype=float32)>

In [20]:
tf.zeros([2,2])

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[0., 0.],
       [0., 0.]], dtype=float32)>

In [95]:
b = tf.ones(shape=(3, 2))

In [96]:
b.ndim

2

In [27]:
import numpy as np
numpy_A = np.arange(1,25, dtype= np.int32)

# X = tf.constant(some_matrix) # capital for matrix or tensor
# Y = tf.constant(vector) # non-capital for vector


In [28]:
numpy_A

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24])

In [29]:
A =tf.constant(numpy_A, shape=(2,3,4))
B =tf.constant(numpy_A)

A,B

(<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
 array([[[ 1,  2,  3,  4],
         [ 5,  6,  7,  8],
         [ 9, 10, 11, 12]],
 
        [[13, 14, 15, 16],
         [17, 18, 19, 20],
         [21, 22, 23, 24]]])>,
 <tf.Tensor: shape=(24,), dtype=int32, numpy=
 array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23, 24])>)

In [9]:
tf.zeros(shape=(2,3), dtype= tf.float16)

<tf.Tensor: shape=(2, 3), dtype=float16, numpy=
array([[0., 0., 0.],
       [0., 0., 0.]], dtype=float16)>

In [22]:
a = tf.zeros([2, 3, 4, 5])
a

<tf.Tensor: shape=(2, 3, 4, 5), dtype=float32, numpy=
array([[[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]],


       [[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]]], dtype=float32)>

In [37]:
A

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]]])>

In [35]:
A[:1]

<tf.Tensor: shape=(1, 3, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]]])>

In [38]:
A[:2,:2]

<tf.Tensor: shape=(2, 2, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20]]])>

In [39]:
A[:2,:2,:2]

<tf.Tensor: shape=(2, 2, 2), dtype=int32, numpy=
array([[[ 1,  2],
        [ 5,  6]],

       [[13, 14],
        [17, 18]]])>

In [41]:
A[:-1,:-1,:-1]

<tf.Tensor: shape=(1, 2, 3), dtype=int32, numpy=
array([[[1, 2, 3],
        [5, 6, 7]]])>

In [45]:
A[:,:,-2:]

<tf.Tensor: shape=(2, 3, 2), dtype=int32, numpy=
array([[[ 3,  4],
        [ 7,  8],
        [11, 12]],

       [[15, 16],
        [19, 20],
        [23, 24]]])>

In [3]:
r4tensor = tf.zeros([2, 3, 4, 5])
r4tensor

<tf.Tensor: shape=(2, 3, 4, 5), dtype=float32, numpy=
array([[[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]],


       [[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]]], dtype=float32)>

In [11]:
r2tensor = tf.constant([[10,7],
                       [3,4]])
r2tensor.shape

TensorShape([2, 2])

In [12]:
r2tensor.ndim

2

In [8]:
r2tensor[:,-1]

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([7, 4])>

In [13]:
r3tensor = r2tensor[...,tf.newaxis]
r3tensor

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]])>

In [14]:
r3tensor = r2tensor[:,:,tf.newaxis]
r3tensor

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]])>

In [15]:
tf.expand_dims(r2tensor, axis= -1)


<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]])>

In [16]:
tf.expand_dims(r2tensor, axis= 0)

<tf.Tensor: shape=(1, 2, 2), dtype=int32, numpy=
array([[[10,  7],
        [ 3,  4]]])>

In [17]:
some_list = [1,2,3,4]
some_list[:2]

[1, 2]

In [30]:
A

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]]])>

In [35]:
A[:,:1,:2]

<tf.Tensor: shape=(2, 1, 2), dtype=int32, numpy=
array([[[ 1,  2]],

       [[13, 14]]])>

# Manipulating tensors (tensor operations)

## basic operations

In [36]:
A + 10

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[11, 12, 13, 14],
        [15, 16, 17, 18],
        [19, 20, 21, 22]],

       [[23, 24, 25, 26],
        [27, 28, 29, 30],
        [31, 32, 33, 34]]])>

In [38]:
C = A + 10

In [39]:
C

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[11, 12, 13, 14],
        [15, 16, 17, 18],
        [19, 20, 21, 22]],

       [[23, 24, 25, 26],
        [27, 28, 29, 30],
        [31, 32, 33, 34]]])>

In [40]:
A

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]]])>

In [41]:
A*2

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 2,  4,  6,  8],
        [10, 12, 14, 16],
        [18, 20, 22, 24]],

       [[26, 28, 30, 32],
        [34, 36, 38, 40],
        [42, 44, 46, 48]]])>

In [42]:
type(A)

tensorflow.python.framework.ops.EagerTensor

### built-in TF function

In [45]:
C = tf.add(A,8)
C

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 9, 10, 11, 12],
        [13, 14, 15, 16],
        [17, 18, 19, 20]],

       [[21, 22, 23, 24],
        [25, 26, 27, 28],
        [29, 30, 31, 32]]])>

In [57]:
%%timeit
C = tf.add(A,8)

137 µs ± 9.19 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [56]:
A_numpy =np.arange(1,25,1)
A_numpy = np.reshape(A_numpy,(2,3,4))
type(A_numpy)

numpy.ndarray

In [51]:
A_numpy

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [58]:
%%timeit
C_numpy = np.add(A,8)

14.2 µs ± 1.03 µs per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


https://realpython.com/numpy-tensorflow-performance/#conclusion

While the NumPy example proved quicker by a hair than TensorFlow in this case, it’s important to note that TensorFlow really shines for more complex cases. With our relatively elementary regression problem, using TensorFlow arguably amounts to “using a sledgehammer to crack a nut,” as the saying goes.

## matrix multiplication

dot matrix not element wide
dot product

# changing datatype of tensor

reduce memory in RAM, but reduce calculation accuracy

In [65]:
B = tf.constant([1.7,7.4])
B

<tf.Tensor: shape=(2,), dtype=float32, numpy=array([1.7, 7.4], dtype=float32)>

In [66]:
C = tf.constant([7,10])
C

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ 7, 10])>

In [67]:
D = tf.cast(B, dtype=tf.float16)
D

<tf.Tensor: shape=(2,), dtype=float16, numpy=array([1.7, 7.4], dtype=float16)>

# tensor aggregation

In [68]:
D = tf.constant([-7,-10])
D

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ -7, -10])>

In [69]:
tf.abs(D)

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ 7, 10])>

### aggregation
- minimum
- maximum
- mean
- sum

In [80]:
A = tf.constant(np.random.randint(0,100,size=50))
A

<tf.Tensor: shape=(50,), dtype=int32, numpy=
array([15, 79, 56, 95, 30, 75, 93, 50,  7, 95, 90, 67, 69, 51, 34, 54, 16,
       21, 55, 38, 95, 19, 84, 87, 95, 30, 54, 12, 57, 58, 36, 26, 63, 17,
       62, 36, 36, 65, 42, 14, 74, 32, 96, 84, 64, 24, 62, 97, 26, 12])>

In [81]:
tf.size(A)

<tf.Tensor: shape=(), dtype=int32, numpy=50>

In [86]:
B =tf.reshape(A,(5,-1))
B

<tf.Tensor: shape=(5, 10), dtype=int32, numpy=
array([[15, 79, 56, 95, 30, 75, 93, 50,  7, 95],
       [90, 67, 69, 51, 34, 54, 16, 21, 55, 38],
       [95, 19, 84, 87, 95, 30, 54, 12, 57, 58],
       [36, 26, 63, 17, 62, 36, 36, 65, 42, 14],
       [74, 32, 96, 84, 64, 24, 62, 97, 26, 12]])>

In [92]:
tf.reduce_max(B), tf.reduce_min(B), tf.reduce_mean(B)

(<tf.Tensor: shape=(), dtype=int32, numpy=97>,
 <tf.Tensor: shape=(), dtype=int32, numpy=7>,
 <tf.Tensor: shape=(), dtype=int32, numpy=52>)

In [93]:
tf.math.reduce_mean(B)

<tf.Tensor: shape=(), dtype=int32, numpy=52>

*** tf cannot return std due to unfit of original dtype

In [105]:
B = tf.cast(B, dtype=tf.float32)

In [106]:
tf.math.reduce_std(B).dtype

tf.float32

In [107]:
tf.math.reduce_std(B)

<tf.Tensor: shape=(), dtype=float32, numpy=27.639458>

In [108]:
tf.math.reduce_std(B).numpy()

27.639458

https://www.tensorflow.org/probability/install

Note: Since TensorFlow is not included as a dependency of the TensorFlow Probability package (in setup.py), you must explicitly install the TensorFlow package (tensorflow or tensorflow-gpu). This allows us to maintain one package instead of separate packages for CPU and GPU-enabled TensorFlow.

In [115]:
# !pip install tensorflow-probability==0.18.0

   ---------------------------------------- 0.0/6.6 MB ? eta -:--:--
   ---------------------------------------- 0.0/6.6 MB 640.0 kB/s eta 0:00:11
   -- ------------------------------------- 0.3/6.6 MB 4.2 MB/s eta 0:00:02
   ----- ---------------------------------- 1.0/6.6 MB 7.7 MB/s eta 0:00:01
   ---------- ----------------------------- 1.7/6.6 MB 9.8 MB/s eta 0:00:01
   --------------- ------------------------ 2.6/6.6 MB 11.7 MB/s eta 0:00:01
   ---------------------- ----------------- 3.7/6.6 MB 13.9 MB/s eta 0:00:01
   ------------------------ --------------- 4.1/6.6 MB 13.0 MB/s eta 0:00:01
   ------------------------------- -------- 5.2/6.6 MB 15.0 MB/s eta 0:00:01
   ------------------------------- -------- 5.2/6.6 MB 14.5 MB/s eta 0:00:01
   ------------------------------- -------- 5.2/6.6 MB 14.5 MB/s eta 0:00:01
   ------------------------------- -------- 5.2/6.6 MB 14.5 MB/s eta 0:00:01
   ------------------------------- -------- 5.2/6.6 MB 14.5 MB/s eta 0:00:01
   ------

ImportError: This version of TensorFlow Probability requires TensorFlow version >= 2.16; Detected an installation of version 2.10.1. Please upgrade TensorFlow to proceed.

This is the 0.18.0 release of TensorFlow Probability. It is tested and stable against TensorFlow version 2.10 and JAX 0.3.17 .
https://github.com/tensorflow/probability/releases

In [116]:
import tensorflow_probability as tfp

In [118]:
B_int =tf.reshape(A,(5,-1))
B_int

<tf.Tensor: shape=(5, 10), dtype=int32, numpy=
array([[15, 79, 56, 95, 30, 75, 93, 50,  7, 95],
       [90, 67, 69, 51, 34, 54, 16, 21, 55, 38],
       [95, 19, 84, 87, 95, 30, 54, 12, 57, 58],
       [36, 26, 63, 17, 62, 36, 36, 65, 42, 14],
       [74, 32, 96, 84, 64, 24, 62, 97, 26, 12]])>

In [119]:
tfp.stats.variance(B_int)

<tf.Tensor: shape=(10,), dtype=int32, numpy=array([980, 569, 211, 846, 555, 344, 667, 942, 354, 951])>

In [120]:
tfp.stats.variance(B)

<tf.Tensor: shape=(10,), dtype=float32, numpy=
array([980.4    , 569.04   , 210.63998, 845.75995, 555.2    , 344.16   ,
       667.36   , 942.8    , 353.84   , 951.0399 ], dtype=float32)>